src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -3152,14 +3152,16 @@ void readInnerClasses(ClassSymbol c) {
                 if (name == names.empty)
                     name = names.one;
                 ClassSymbol member = enterClass(name, outer);
-                if ((flags & STATIC) == 0) {
-                    ((ClassType)member.type).setEnclosingType(outer.type);
-                    if (member.erasure_field != null)
-                        ((ClassType)member.erasure_field).setEnclosingType(types.erasure(outer.type));
-                }
-                if (c == outer && member.owner == c) {
-                    member.flags_field = flags;
-                    enterMember(c, member);
+                if ((member.flags_field & FROM_SOURCE) == 0) {
```

Comment:
I agree with the fix -- I guess I'd slightly prefer to see a warning if a mismatch is detected, but I'll leave that to you.